targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/bsp/rng/adi_rng.c

```diff
@@ -0,0 +1,796 @@
+ *              - #ADI_RNG_INVALID_STATE[D]       Random number ready status is not set
+                - #ADI_RNG_INVALID_PARAM [D]          Argument is incorrect.
+ *
+ * @sa        adi_rng_Open().
+ * @sa        adi_Rng_RegisterCallback().
+ */
+ADI_RNG_RESULT adi_rng_GetOscDiff (
+                                   ADI_RNG_HANDLE const hDevice,
+                                   uint32_t       const nIndex,
+                                   uint8_t*       const pOscDiff
+                                   )
+{
+    ADI_RNG_DEV_TYPE *pDevice = (ADI_RNG_DEV_TYPE*)hDevice;
+
+#ifdef ADI_DEBUG
+   if (ADI_RNG_INVALID_HANDLE(pDevice)){
+        return ADI_RNG_BAD_DEV_HANDLE;
+    }
+
+    if (NULL == pDevice->pData) {
+        return ADI_RNG_NOT_INITIALIZED;
+    }
+
+    if ((NULL == pOscDiff) || (nIndex > 3u)) {
+        return( ADI_RNG_INVALID_PARAM );
+    }
+
+    if ((pDevice->pRNG->STAT & BITM_RNG_STAT_RNRDY) == 0u) {
+        return ADI_RNG_INVALID_STATE;
+    }
+#endif
+
+    /* Get the Osc Difference Register */
+    *pOscDiff = (uint8_t)pDevice->pRNG->OSCDIFF[nIndex];
+
+    return ADI_RNG_SUCCESS;
+}
+
+/*!
+ * @brief  Retrieve the current RNG sample length prescale and reload value configured in the device.
+ *
+ * @param[in]   hDevice        Device handle obtained from adi_rng_Open().
+ * @param[out]  pLenPrescaler  Pointer to an application-defined variable into which the prescaler value is written.
+ * @param[out]  pLenReload     Pointer to an application-defined variable into which the reload value for the sample counter is written.
+ *
+ * @return      Status
+ *                - #ADI_RNG_SUCCESS                    Call completed successfully.
+ *                - #ADI_RNG_BAD_DEV_HANDLE [D]     Invalid device handle parameter.
+ *                - #ADI_RNG_NOT_INITIALIZED [D]    Device has not been initialized for use, see #adi_rng_Open().
+                  - #ADI_RNG_INVALID_PARAM [D]          Argument is incorrect.
+ *
+ *
+ * @sa        adi_rng_Open().
+ * @sa        adi_rng_RegisterCallback().
+ */
+ADI_RNG_RESULT adi_rng_GetSampleLen (
+                                     ADI_RNG_HANDLE const hDevice,
+                                     uint16_t*      const pLenPrescaler,
+                                     uint16_t*      const pLenReload
+                                     )
+{
+    ADI_RNG_DEV_TYPE *pDevice = (ADI_RNG_DEV_TYPE*)hDevice;
+
+#ifdef ADI_DEBUG
+   if (ADI_RNG_INVALID_HANDLE(pDevice)){
+        return ADI_RNG_BAD_DEV_HANDLE;
+    }
+
+    if (NULL == pDevice->pData) {
+        return ADI_RNG_NOT_INITIALIZED;
+    }
+
+    if ((NULL == pLenPrescaler) || (NULL == pLenReload)) {
+        return ADI_RNG_INVALID_PARAM;
+    }
+#endif
+
+    *pLenPrescaler = (pDevice->pRNG->LEN & BITM_RNG_LEN_PRESCALE) >> BITP_RNG_LEN_PRESCALE;
+    *pLenReload    = (pDevice->pRNG->LEN & BITM_RNG_LEN_RELOAD)   >> BITP_RNG_LEN_RELOAD;
```

Comment:
The `LEN` register is read twice here but there is no lock in place - is it conceivable that the value has changed in between? Perhaps it would be safer to read it into a temporary once and to do the splitting into the prescale and reload values afterwards.

Reply:
The value in LEN register does not change once set. It's a configuration register.